### ISSUE TITLE
Inquiries about Monocular Image Inference Implementation in ImVoxelNet

### ISSUE THREAD

**Issue by henu77** [SamsungLabs/imvoxelnet#85]
# Questions Regarding Model Configuration, Dataset, and Inference

I have thoroughly reviewed all relevant discussions, and my specific scenario appears to be unique and undocumented. I would be grateful if you could provide guidance on the following questions:

## 1. Model Configuration Clarification

In GitHub issue #38, you mentioned that the `imvoxelnet_total_sunrgbd_fast.py` configuration is the only model that can run without LiDAR2Img for the SUN RGB-D Total3DUnderstanding benchmark, using an additional 2D head to predict camera angles. Could you confirm whether my understanding is correct:

 a. This configuration accepts single RGB images as input.

 b. Directly outputs 3D bounding boxes with category information without requiring LiDAR data.

## 2. Dataset and Visualization

a. Could you share the approximate storage requirements for the complete SUN RGB-D dataset?

 b. After downloading the dataset, would running:

```bash
python tools/test.py configs/imvoxelnet/imvoxelnet_total_sunrgbd_fast.py checkpoints/20211007_105247.pth --out results --show
```

produce 2D images with projected 3D bounding boxes similar to those shown in the GitHub repository and paper?

## 3. Single-image Inference Demo
a. The point cloud demo (pcd_demo.py) works elegantly for PCD inputs. Could you provide guidance on adapting this for single RGB image inference?

b. I attempted to modify the inference_demo.ipynb but encountered challenges in handling image inputs. You can view the error message in the attached image.

c. Any suggestions for modifying the data pipeline or visualization would be greatly appreciated.
Thank you for your time and for making your code publicly available. I look forward to your insights.

![Image](https://github.com/user-attachments/assets/1e85903e-0ada-467b-aaea-9b50b5361cd7)

**Comment by filaPro** [comment]
Hi @henu77 ,

1a. Single image and intrinsic camera matrix, i think. Extrinsic matrix will be predicted by additional head.
1b. We don't use lidar or depth in this paper.
2a. May be ~5GB.
2b. Should be true.
3. Actually ImVoxelNet is supported in mmdetection3d. And they provide a [script](https://github.com/open-mmlab/mmdetection3d/blob/main/demo/mono_det_demo.py) for running on a single image. But mmdetection3d project is not supported any more, so not sure if it is easy to run this script successfully.

**Comment by henu77** [comment]
Thank you for your detailed explanation. I have successfully implemented 3D box prediction based on a single image, but I am encountering challenges when trying to visualize the 3D bounding boxes on the 2D input image.

I referenced your excellent response in issue #55 and followed the inference process using the **20211007_105247.pth** checkpoint and the **imvoxelnet_total_sunrgbd_fast.py** configuration file. During this process, I set up the `lidar2img `parameters as follows:

```
lidar2img = {
    "intrinsic": torch.tensor(
        [[529.5, 0, 365, 0], 
         [0, 529.5, 265, 0], 
         [0, 0, 1, 0], 
         [0, 0, 0, 1]],
        dtype=torch.float32,
    ),
    "origin": np.array([0, 3, -1]),
    "extrinsic": torch.tensor(
        [[1, 0, 0, 0], 
         [0, 1, 0, 0], 
         [0, 0, 1, 0], 
         [0, 0, 0, 1]], dtype=torch.float32
    ),
}
```

The model outputs the following results:
```
{
    'bboxes_3d': torch.Size([1000, 7]),
    'scores_3d': torch.Size([1000]),
    'labels_3d': torch.Size([1000]), 
    'angles': torch.Size([2]), 
    'layout': torch.Size([1, 7])
}
```

When attempting to project the predicted 3D bounding boxes onto the input image, I followed these steps:

First, I passed the predicted `angles `into the [`_compute_projection` ](https://github.com/SamsungLabs/imvoxelnet/blob/655094d44d7c63af79564bab4e62270ac181838f/mmdet3d/models/detectors/imvoxelnet.py#L115)function to compute the `projection`.
Then, I used the [`draw_corners` ](https://github.com/SamsungLabs/imvoxelnet/blob/655094d44d7c63af79564bab4e62270ac181838f/mmdet3d/datasets/dataset_wrappers.py#L88)function to draw the 3D bounding boxes.
However, the visualization results are not as expected. 

![Image](https://github.com/user-attachments/assets/2dc31b87-352f-4abc-ac75-127085756a16)

Below is a summary of the relevant variables and their values:
Intrinsic Matrix :
```
[[529.5,   0. , 365. ],
 [  0. , 529.5, 265. ],
 [  0. ,   0. ,   1. ]]
```
Projection Matrix :
```
[[[ 119.77985      82.44891      -7.578066      0.        ]
  [   5.0627856    41.06457    -127.51828       0.        ]
  [   0.            0.98834264   -0.15224592    0.        ]]]
```
bboxes_3d.corners: 
```
[[ 0.2507319   1.9770359  -1.1872268 ]
 [ 0.2507319   1.9770359   0.27221036]
 [-1.3158224   2.9879255   0.27221036]
 [-1.3158224   2.9879255  -1.1872268 ]
 [ 1.4221265   3.7923217  -1.1872268 ]
 [ 1.4221265   3.7923217   0.27221036]
 [-0.14442773  4.803211    0.27221036]
 [-0.14442773  4.803211   -1.1872268 ]]
```
label:
```
bed
```

Despite following these steps, the projected 3D bounding boxes do not align correctly with the objects in the image. Could you provide guidance on how to properly visualize the 3D boxes on the 2D image? Specifically, I would appreciate clarification on:

- Whether the projection matrix is being computed correctly.
- How to ensure the corners of the 3D bounding box are accurately mapped to the 2D image plane.
- Any additional preprocessing or postprocessing steps that might be necessary for correct visualization.

Thank you for your assistance!

**Comment by filaPro** [comment]
Were you able to run our sunrgbd preprocessing in total3d format? And than run our visualization script on the same scene?

**Comment by henu77** [comment]
> Were you able to run our sunrgbd preprocessing in total3d format? And than run our visualization script on the same scene?

I seem to have identified an issue in my code. I did not utilize the sunrgbd processing you provided to convert the input image into the total3d format. Instead, I merely loaded the input image using OpenCV, transformed it into a PyTorch tensor, and applied standard normalization procedures.

Should I wish to perform inference on images I have captured myself, what steps should I take to process these images prior to feeding them into the model?

**Comment by filaPro** [comment]
Also be careful in these 3 [lines](https://github.com/SamsungLabs/imvoxelnet/blob/master/mmdet3d/models/detectors/imvoxelnet.py#L117) as after resize intrinsic matrix is changed.

**Comment by henu77** [comment]
Thank you very much for your guidance. I will revise my code and try again tomorrow.

**Comment by henu77** [comment]
My steps are as follows:
1. Load the model
2. 
3.

**Comment by henu77** [comment]
My steps are as follows:
1. Load the model
``` python
from mmdet3d.apis.inference import init_model, inference_detector
from mmengine.runner import load_checkpoint
from mmdet3d.models.detectors import ImVoxelNet_v2
import torch
import matplotlib.pyplot as plt

from checkpoints.imvoxelnet_total_sunrgbd_fast import (
    model,
    test_cfg,
    train_cfg,
    class_names,
    dataset_type,
    data_root,
    train_pipeline,
    test_pipeline,
    metainfo,
    backend_args,
)

model = ImVoxelNet_v2(
    backbone=model["backbone"],
    neck=model["neck"],
    neck_3d=model["neck_3d"],
    bbox_head=model["bbox_head"],
    n_voxels=model["n_voxels"],
    voxel_size=model["voxel_size"],
    head_2d=model["head_2d"],
    train_cfg=train_cfg,
    test_cfg=test_cfg,
)

checkpoint = r"checkpoints\20211007_105247.pth"

checkpoint = load_checkpoint(model, checkpoint, map_location="cpu")
# save the dataset_meta in the model for convenience
if "dataset_meta" in checkpoint.get("meta", {}):
    # mmdet3d 1.x
    model.dataset_meta = checkpoint["meta"]["dataset_meta"]
elif "CLASSES" in checkpoint.get("meta", {}):
    # < mmdet3d 1.x
    classes = checkpoint["meta"]["CLASSES"]
    model.dataset_meta = {"classes": classes}

    if "PALETTE" in checkpoint.get("meta", {}):  # 3D Segmentor
        model.dataset_meta["palette"] = checkpoint["meta"]["PALETTE"]
else:
    # < mmdet3d 1.x
    model.dataset_meta = {"classes": class_names}

    if "PALETTE" in checkpoint.get("meta", {}):  # 3D Segmentor
        model.dataset_meta["palette"] = checkpoint["meta"]["PALETTE"]
```
2. Process the input image
``` python
from mmengine.registry import TRANSFORMS
from mmdet3d.datasets import MultiViewPipeline
from mmcv.transforms import Compose
import numpy as np

TRANSFORMS.register_module(name='MultiViewPipeline', module=MultiViewPipeline)

transform = Compose(test_pipeline)

inputs ={
        "img_info": [{"filename": "checkpoints\sunrgbd_0000103.jpg"}],
        "img_prefix": [
            "D:\Project\mmdetection3d\\"
        ],
        "lidar2img": {
            "extrinsic": [
                np.array([[1, 0, 0, 0], [0, 1, 0, 0], [0, 0, 1, 0], [0, 0, 0, 1]])
            ],
            "intrinsic": [
                np.array(
                    [[529.5, 0, 365, 0], [0, 529.5, 265, 0], [0, 0, 1, 0], [0, 0, 0, 1]]
                )
            ],
            "origin": [np.array([0, 3, -1])]
        },
        
    }

input_data = transform(inputs)
```
3. Use the `simple_test` function in the model to predict the 3D box.
``` python

from mmdet3d.structures import DepthInstance3DBoxes

lidar2img = {
    "intrinsic": torch.tensor(
        [[529.5, 0, 365, 0], 
         [0, 529.5, 265, 0], 
         [0, 0, 1, 0], 
         [0, 0, 0, 1]],
        dtype=torch.float32,
    ),
    "origin": np.array([0, 3, -1]),
    "extrinsic": torch.tensor(
        [[1, 0, 0, 0], 
         [0, 1, 0, 0], 
         [0, 0, 1, 0], 
         [0, 0, 0, 1]], dtype=torch.float32
    ),
}
img_metas = [
    {   
        "filename": image_path,
        "img_shape": img_shape,
        "ori_shape": ori_shape,
        "lidar2img": lidar2img,
        "box_type_3d": DepthInstance3DBoxes,
    }
]

result = model.simple_test(input_data['inputs']['img'].cuda(), img_metas)
```

4. Use [`_compute_projection(img_meta, stride, angles)`](https://github.com/SamsungLabs/imvoxelnet/blob/655094d44d7c63af79564bab4e62270ac181838f/mmdet3d/models/detectors/imvoxelnet.py#L115) to calculate the projection

   For the three parameters `img_meta`, `stride`, and `angles`, I used the same `img_meta `as in step 3, set `stride `to 4, and used the result obtained from step 3 as `angles`.

5. Use [`draw_corners(img, corners, color, projection)` ](https://github.com/SamsungLabs/imvoxelnet/blob/655094d44d7c63af79564bab4e62270ac181838f/mmdet3d/datasets/dataset_wrappers.py#L88)to draw the 3D box
   For these parameters, `img `is read using `skimage.io.imread(image_path)`, `corners `are the results output by the model in the step 3, `color `is set to red, and `projection` comes from the results of the step 4.


**Can you help me check what problems exist in these steps?**
